design/thrift-proxy-design.md

```diff
@@ -0,0 +1,215 @@
+
+## Thrift Proxy
+`extensions.filters.network.thrift_proxy.v3.ThriftProxy`  
+The Thrift Proxy, like any Envoy proxy, defines various configuration elements, including route configurations, filters, access logs, and the Thrift route discovery service. 
+Each instance of the Thrift Proxy corresponds to one Thrift RouteConfiguration configuration.
+
+## Thrift RouteConfiguration
+`extensions.filters.network.thrift_proxy.v3.RouteConfiguration`  
+The Thrift RouteConfiguration is responsible for defining routes and their corresponding rules within the Thrift filter. It can be identified by a unique name used for route discovery. 
+Each RouteConfiguration includes an array of Thrift routes.
+
+## Thrift Routes
+Thrift Routes are defined within RouteConfigurations and represent specific routing rules for Thrift traffic. 
+Each Thrift route consists of two main entries: the Thrift `RouteMatch` and the Thrift `RouteAction`.
+
+### Thrift RouteMatch
+`extensions.filters.network.thrift_proxy.v3.RouteMatch`  
+Thrift RouteMatches are essential for defining matching criteria for Thrift routes. 
+Each Thrift route **must** have a matching rule, which can be based on either the Thrift `service_name` or the Thrift `method_name`, but not both. 
+Additionally, the `RouteMatch` supports an array of `config.route.v3.HeaderMatcher` for additional matching conditions.
+
+### Thrift RouteAction
+`extensions.filters.network.thrift_proxy.v3.RouteAction`  
+Thrift RouteActions specify the Thrift cluster responsible for handling incoming requests. 
+They offer advanced features like weighted clusters, request mirror policies, and rate limits.
+
+## Thrift Route Discovery Service(TRDS)
+`extensions.filters.network.thrift_proxy.v3.Trds`  
+The Thrift Route Discovery Service (TRDS) serves as a critical component for Thrift route discovery. 
+It is currently compatible with Aggregated Service Discovery (ADS) and is responsible for fetching specific route configurations. 
+TRDS identifies the desired route configuration by name, allowing for the use of different configurations as needed.
+
+# High-Level Design
+
+The implementation is organized into four essential areas:
+### ThriftProxy CRD
+The foundation of our design lies in the creation of the ThriftProxy CRD, empowering users to define intricate routing rules and specify the services they require. 
+By introducing this Custom Resource Definition, we provide a straightforward and flexible means for users to configure Thrift routing within Contour.
+### Thrift Listener
+A dedicated Thrift Listener is responsible for receiving incoming Thrift traffic on a distinct port separate from gRPC and HTTP protocols.
+### Thrift RouteConfiguration
+Thrift RouteConfiguration defines the Envoy configurations that correspond to user-defined routing rules.
+### Aggregated Service Discovery
+Aggregated Service Discovery (ADS) to enable Thrift route discovery.
+
+# Detailed Design
+
+## ThriftProxy CRD
```

Comment:
Would it be possible for you all to use Gateway API and a new *Route CRD instead? (i.e. ThriftRoute CRD)

I say this for a couple reasons, but mainly because if we expect a unique Envoy Listener specific to Thrift Proxy routes (i.e. not the existing two Listeners on plain HTTP or HTTPS ports), we need to leverage Contour's dynamic Listener support which we only really intended to support with Gateway API.

We should probably add some discussion to this design about Envoy Listeners, how we expect Thrift Proxy support to interact with them.

Reply:
Thanks for pointing out the Gateway API point, and I agree with you on adding ThriftRoute CRD for the listeners.

I haven't gotten my hands dirty with the Gateway API stuff, and we don't use it in my company yet. That is why I tried to be explicit about not having it as a Goal of this design doc.

However, I can over 2 options here:
- Open an issue to track adding support for ThriftRoute to Gateway API
- I can include it in this design if we agree that it is important and we should cover it in early stages.

Reply:
I think this is still an open point to add to this doc, the above sections mention adding a dedicated Thrift Listener but would be good to add some more detail to this Detailed Design section on this topic (how will the port used for Thrift be configured, what does the filter chain/match look like, etc.)